File: pkgs/oizys/src/oizys/github.nim
import std/[httpclient,logging, os, strformat, strutils, json, tables, tempfiles, times, strtabs]
import jsony, hwylterm, hwylterm/logging, zippy/ziparchives, resultz
import ./[exec, context]


template withTmpDir(body: untyped): untyped =
  let tmpDir {.inject.} = createTempDir("oizys","")
  body
  removeDir tmpDir

type
  GhArtifact = object
    id: int
    name: string
    url: string
    archive_download_url*: string

  GhWorkflowRun = object
    id*: int
    node_id: string
    run_number: int
    event: string
    status: string
    conclusion: string
    html_url: string
    workflow_id: int
    created_at*: string # use datetime?
    updated_at: string # use datetime?

  ListGhArtifactResponse = object
    total_count: int
    artifacts: seq[GhArtifact]

  ListGhWorkflowResponse = object
    total_count: int
    workflow_runs: seq[GhWorkflowRun]

var ghToken = getEnv "GITHUB_TOKEN"

proc checkToken() {.inline.} =
  if ghToken == "": fatalQuit "GITHUB_TOKEN not set"

proc ghClient(
  maxRedirects = 5
): HttpClient =
  checkToken()
  result = newHttpClient(maxRedirects = maxRedirects)
  result.headers = newHttpHeaders({
    "Accept"              : "application/vnd.github+json",
    "Authorization"       : fmt"Bearer {ghToken}",
    "X-GitHub-Api-Version": "2022-11-28",
  })


proc getGhApi(url: string): Response =
  let client = ghClient()
  try:
    result = client.get(url)
  except:
    error getCurrentExceptionMsg()
    error fmt"github api request failed: {url}"
    if result != nil:
      error fmt"response: {result.body}"
    quit QuitFailure
  finally:
    close client

proc postGhApi(url: string, body: JsonNode) =
  checkToken()
  let client = newHttpClient()
  client.headers = newHttpHeaders({
    "Accept"              : "application/vnd.github+json",
    "Authorization"       : fmt"Bearer {ghToken}",
    "X-GitHub-Api-Version": "2022-11-28",
  })
  var response: Response
  try:
    response = client.post(url, body = $body)
    info fmt"Status: {response.code}"
  except:
    errorQuit "failed to get response code"
  finally:
    close client
  if response.code != Http204:
    errorQuit "failed to post github api request"


proc getInProgressRun(
  workflow: string,
  timeout: int = 10000
): Opt[GhWorkflowRun] =
  ## wait up to 10 seconds to try to fetch ongoing run url
  let
    start = now()
    timeoutDuration = initDuration(milliseconds = timeout)

  withSpinner(fmt"waiting for {workflow} workflow to start"):
    while (now() - start) < timeoutDuration:
      let response = getGhApi(fmt"https://api.github.com/repos/daylinmorgan/oizys/actions/workflows/{workflow}/runs")
      let runs = fromJson(response.body,  ListGhWorkflowResponse).workflow_runs
      if runs[0].status in ["in_progress", "queued"]:
        return ok runs[0]
      sleep 500

  warn "timeout reached waiting for workflow to start"

proc `%`*(table: StringTableRef): JsonNode =
  ## Generic constructor for JSON data. Creates a new `JObject JsonNode`.
  result = newJObject()
  for k, v in table: result[k] = %v

proc createDispatch*(workflowFileName: string, `ref`: string, inputs: StringTableRef) =
  ## https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event
  let workflow =
    if workflowFileName.endsWith(".yml") or workflowFileName.endsWith(".yaml"): workflowFileName
    else: workflowFileName & ".yml"
  let body = %*{"ref": `ref`, "inputs": inputs}
  info fmt"creating dispatch event for {workflow}"
  debug "with body: " & $body
  postGhApi(
   fmt"https://api.github.com/repos/daylinmorgan/oizys/actions/workflows/{workflow}/dispatches",
   body
  )
  case getInProgressRun(workflow)
  of Some(run):
    info "view workflow run at: " & run.html_url
  of None:
    warn "couldn't determine workflow url"

proc listUpdateRuns(): seq[GhWorkflowRun] =
  ## get update.yml runs
  ## endpoint https://api.github.com/repos/OWNER/REPO/actions/workflows/WORKFLOW_ID/runs
  debug "listing update workflows"
  let response = getGhApi("https://api.github.com/repos/daylinmorgan/oizys/actions/workflows/update.yml/runs")
  fromJson(response.body,  ListGhWorkflowResponse).workflow_runs

proc getLastUpdateRun*():  GhWorkflowRun =
  let runs = listUpdateRuns()
  let run = runs[0]
  if run.conclusion == "failure":
    fatalQuit bb(fmt("Most recent run was not successful\n[b]runID[/]: {run.id}\n[b]conclusion[/]: {run.conclusion}"))
  if run.status in ["in_progress", "queued"]:
    fatalQuit bb(fmt("Most recent run is not finished\nview workflow run at: {run.html_url}"))
  result = run


proc getArtifacts(runId: int): seq[GhArtifact] =
  ## get workflow artifacts
  ## https://api.github.com/repos/OWNER/REPO/actions/runs/RUN_ID/artifacts
  let response = getGhApi(fmt"https://api.github.com/repos/daylinmorgan/oizys/actions/runs/{runId}/artifacts")
  fromJson(response.body, ListGhArtifactResponse).artifacts

proc getUpdateSummaryArtifact(runId: int): GhArtifact =
  let name = "summary"
  let artifacts = getArtifacts(runId)
  for artifact in artifacts:
    if artifact.name == name:
      return artifact
  fatalQuit fmt"failed to find summary for run id: {runID}"

proc getUpdateSummaryUrl(runID: int): string =
  ## https://api.github.com/repos/OWNER/REPO/actions/artifacts/ARTIFACT_ID/ARCHIVE_FORMAT
  let artifact = getUpdateSummaryArtifact(runID)
  # httpclient was forwarding the Authorization headers,
  # which confused Azure where the archive lives...
  var response: Response
  try:
    let client = ghClient(maxRedirects = 0)
    response = client.get(artifact.archive_download_url)
  except:
    errorQuit fmt("fetching summary failed:\n\n{response.headers}\n\n{response.body}")

  if "location" notin response.headers.table:
    errorQuit fmt("fetching summary failed:\n\n{response.headers}\n\n{response.body}")

  let location = response.headers.table.getOrDefault("location", @[])
  if location.len == 0: errorQuit fmt("location header missing url?")
  return location[0]

proc fetchUpdateSummaryFromUrl(url: string, host: string): string =
  withTmpDir:
    let client = newHttpClient()
    client.downloadFile(url, tmpDir / "summary.zip")
    let reader = openZipArchive(tmpDir / "summary.zip")
    try:
      result = reader.extractFile(host & "-summary.md")
    finally:
      reader.close()

proc getUpdateSummary*(runId: int, host: string): string =
  withSpinner("fetching update summary"):
    let url = getUpdateSummaryUrl(runId)
    result = fetchUpdateSummaryFromUrl(url, host)

type
  GitRepo = object
    path: string

proc git(r: GitRepo, rest: varargs[string]): Command =
  result.exe = "git"
  result.addArgs ["-C", r.path]
  result.addArgs rest

proc checkGit(code: int) =
  if code != 0: fatalQuit "git had a non-zero exit status"

proc fetch(r: GitRepo) =
  let code = r.git("fetch", "origin").run()
  checkGit code

proc status(r: GitRepo) =
  let (output, _, code) =  r.git("status", "--porcelain").runCapt()
  checkGit code
  if output.len > 0:
    info "unstaged commits, cowardly exiting..."
    quit QuitFailure

proc rebase(r: GitRepo, `ref`: string) =
  r.status()
  let code = r.git("rebase", `ref`).run()
  checkGit code

proc updateRepo*() =
  let repo = GitRepo(path: getFlake())
  fetch repo
  rebase(repo, "origin/flake-lock")

type
  GhPullResponse = object
    id*: int
    state: string
    title: string
    merged: bool
    merge_commit_sha: string

  GhCommit = object
    sha, url: string

  GhBranch = object
    name: string
    commit: GhCommit
    protected: bool

proc getGhPull*(owner: string, repo: string, number: int): GhPullResponse =
  fromJson(
    getGhApi(
      fmt"https://api.github.com/repos/{owner}/{repo}/pulls/{number}"
    ).body,
    typeof(result)
  )


proc getGhBranches*(owner: string, repo: string): seq[GhBranch] =
  # NOTE: there are pages...
  fromJson(
    getGhApi(
      fmt"https://api.github.com/repos/{owner}/{repo}/branches"
    ).body,
    typeof(result)
  )


proc getGhBranch(owner: string, repo: string, branch: string): GhBranch =
  fromJson(
    getGhapi(
      fmt"https://api.github.com/repos/{owner}/{repo}/branches/{branch}"
    ).body,
    typeof(result)
  )

type
  GhCompare = object
    status: string

proc ghRepoCompare(owner: string, repo: string, base: string, head: string): GhCompare =
  fromJson(
    getGhApi(
      fmt"https://api.github.com/repos/{owner}/{repo}/compare/{base}...{head}"
    ).body,
    typeof(result)
  )
  
  # "status": "behind",

proc prInBranch(owner: string, repo: string, branch: GhBranch,  pr: GhPullResponse): bool =
  if not pr.merged:
    return false

  let compare = ghRepoCompare(owner, repo, pr.merge_commit_sha, branch.commit.sha)

  case compare.status:
  of "behind", "identical":
    return false
  of "ahead":
    return true
  of "diverged":
    fatalQuit "compared status of 'diverged' is not yet supported"
  else:
    fatalQuit "unknown status: " & compare.status


type
  NixpkgsPrStatus* = object
    number: int
    merged: bool
    nixpkgsUnstable: bool
    nixosUnstableSmall: bool
    nixosUnstable: bool

proc getNixpkgsPrStatus*(number: int): NixpkgsPrStatus =
  const
    owner = "nixos"
    repo  = "nixpkgs"

  withSpinner(fmt"checking on status of PR [b]#{number}"):

    let pr = getGhPull(owner, repo, number)
    result.number = number
    result.merged = pr.merged

    if result.merged:
      result.nixpkgsUnstable = prInBranch(owner, repo, getGhBranch(owner, repo, "nixpkgs-unstable"),pr)
      result.nixosUnstableSmall = prInBranch(owner, repo, getGhBranch(owner, repo, "nixos-unstable-small"),pr)
      result.nixosUnstable = prInBranch(owner, repo, getGhBranch(owner, repo, "nixos-unstable"), pr)

proc bb*(status: NixpkgsPrStatus): BbString =
  var s: string

  template mark(cond: bool): untyped =
    if cond: "[green]✓[/]" else: "[red]✗[/]"

  s.add fmt"[b]PR #{status.number}[/] status:"

  if not status.merged:
    s.add " not merged"
  else:
    s.add "\n"
    s.add fmt"├──── " & (mark status.nixpkgsUnstable) & " nixpkgs-unstable"
    s.add "\n"
    s.add fmt"├─ " & mark(status.nixosUnstableSmall) & " nixos-unstable-small"
    s.add "\n"
    s.add fmt"╰─ " & mark(status.nixosUnstable) & " nixos-unstable"

  bb(s)

File: pkgs/oizys-rs/src/github.rs
use super::prelude::*;
use chrono::{DateTime, Utc};
use reqwest::{
    blocking::{Client, Response},
    header, StatusCode,
};
use serde::Deserialize;
use serde::{de::DeserializeOwned, Serialize};
use std::time::{Duration, Instant};
use std::{collections::HashMap, thread};
use tracing::{debug, info};

#[derive(Deserialize, Debug)]
struct GithubCommit {
    sha: String,
    // url: String,
}

#[derive(Deserialize, Debug)]
struct GithubBranch {
    // name: String,
    commit: GithubCommit,
    // protected: bool,
}

#[derive(Debug, Deserialize)]
struct GithubPullRequest {
    // id: i64,
    // state: String,
    // title: String,
    merged: bool,
    merge_commit_sha: String,
}

#[derive(Deserialize)]
struct GithubCompare {
    status: String,
}

#[derive(Debug)]
pub enum PrStatus {
    NotMerged,
    Merged {
        nixpkgs_unstable: bool,
        nixos_unstable_small: bool,
        nixos_unstable: bool,
    },
}

fn style_branch(name: &str, merged: &bool) -> String {
    let branch = style(name);
    let branch = if *merged {
        branch.green()
    } else {
        branch.red()
    };
    branch.to_string()
}
impl std::fmt::Display for PrStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Self::NotMerged => writeln!(f, "{}", style("not merged").red())?,
            Self::Merged {
                nixpkgs_unstable,
                nixos_unstable_small,
                nixos_unstable,
            } => {
                writeln!(
                    f,
                    "-> {}",
                    style_branch("nixpkgs-unstable", nixpkgs_unstable)
                )?;
                writeln!(
                    f,
                    "-> {}",
                    style_branch("nixos-unstable-small", nixos_unstable_small)
                )?;
                writeln!(f, "-> {}", style_branch("nixos-unstable", nixos_unstable))?;
            }
        }

        Ok(())
    }
}

// one of these attributes was causing a parser error?
#[derive(Deserialize, Debug, Clone)]
struct GithubWorkflowRun {
    id: i64,
    //node_id: String,
    //run_number: i64,
    //event: String,
    status: String,
    //conclusion: String,
    //html_url: String,
    //workflow_id: i64,
    created_at: DateTime<Utc>,
    //updated_at: String,
}

#[derive(Deserialize, Debug)]
struct ListGithubWorflowRun {
    //    total_count: i64,
    workflow_runs: Vec<GithubWorkflowRun>,
}

static APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);

fn handle_response<T: DeserializeOwned>(resp: Response) -> Result<T> {
    match resp.status() {
        StatusCode::OK => {
            let data: T = resp.json()?;
            return Ok(data);
        }
        s => bail!(
            "github api had an unexpected error: {s:?}, see below:\n{}",
            resp.text()?
        ),
    };
}

struct GitHubRepo {
    client: Client,
    owner: String,
    repo: String,
}

impl GitHubRepo {
    fn new_client() -> Result<Client> {
        let mut headers = header::HeaderMap::new();
        if let Ok(token) = std::env::var("GITHUB_TOKEN") {
            debug!("using GITHUB_TOKEN for API requests");
            let token = "Bearer ".to_string() + &token;
            let mut token = header::HeaderValue::from_str(&token)?;
            token.set_sensitive(true);
            headers.insert("Authorization", token);
        } else {
            info!("set env var GITHUB_TOKEN to use authenticated requests")
        }
        headers.insert(
            "X-GitHub-Api-Version",
            header::HeaderValue::from_static("2022-11-28"),
        );
        headers.insert(
            "Accept",
            header::HeaderValue::from_static("application/vnd.github+json"),
        );
        Ok(Client::builder()
            .user_agent(APP_USER_AGENT)
            .default_headers(headers)
            .build()?)
    }
    fn new(owner: &str, repo: &str) -> Result<Self> {
        let client = Self::new_client()?;
        Ok(Self {
            client,
            owner: owner.to_string(),
            repo: repo.to_string(),
        })
    }

    fn get<T: DeserializeOwned>(&self, url: &str) -> Result<T> {
        debug!("GET: {}", url);
        let resp = self.client.get(url).send()?;
        handle_response(resp)
    }

    fn get_branch(&self, branch: &str) -> Result<GithubBranch> {
        self.get(&format!(
            "https://api.github.com/repos/{}/{}/branches/{}",
            self.owner, self.repo, branch
        ))
    }

    fn get_pr(&self, number: i64) -> Result<GithubPullRequest> {
        self.get(&format!(
            "https://api.github.com/repos/{}/{}/pulls/{}",
            self.owner, self.repo, number
        ))
    }

    fn compare_branch_to_pr(&self, base: &str, head: &str) -> Result<GithubCompare> {
        self.get(&format!(
            "https://api.github.com/repos/{}/{}/compare/{}...{}",
            self.owner, self.repo, base, head
        ))
    }
    fn pr_in_branch(&self, branch: &str, pr: &GithubPullRequest) -> Result<bool> {
        if !pr.merged {
            return Ok(false);
        }

        let branch = self.get_branch(branch)?;
        let compare = self.compare_branch_to_pr(&pr.merge_commit_sha, &branch.commit.sha)?;

        match compare.status.as_str() {
            "behind" | "identical" => Ok(false),
            "ahead" => Ok(true),
            "diverged" => bail!("compared status of 'diverged' is not yet supported"),
            s => bail!("unknown status {s}"),
        }
    }

    fn run_action(
        &self,
        workflow: &str,
        ref_name: &str,
        inputs: Vec<(String, String)>,
    ) -> Result<()> {
        let url = format!(
            "https://api.github.com/repos/{}/{}/actions/workflows/{}/dispatches",
            self.owner, self.repo, workflow
        );
        debug!("POST: {}", url);
        let body = GithubWorkflowDispatchBody::new(&ref_name, inputs)?;
        let res = self.client.post(url).json(&body).send()?;
        match res.status() {
            StatusCode::NO_CONTENT => Ok(()),
            StatusCode::NOT_FOUND => {
                bail!(
                    "workflow does not exist, see below for API reponse:\n{:?}",
                    &res
                )
            }
            _ => bail!("possible error submitting post request {:?}", &res),
        }
    }

    fn normalize_workflow(name: &str) -> String {
        if !name.ends_with(".yml") && !name.ends_with(".yaml") {
            name.to_string() + ".yml" // I use .yml only
        } else {
            name.to_string()
        }
    }

    fn list_workflow_runs(&self, workflow: &str) -> Result<ListGithubWorflowRun> {
        self.get(&format!(
            "https://api.github.com/repos/{}/{}/actions/workflows/{}/runs",
            self.owner, self.repo, workflow
        ))
    }

    fn is_current_run(run: &GithubWorkflowRun, now: &DateTime<Utc>) -> bool {
        ["in_progress", "queued"].contains(&run.status.as_str()) && run.created_at > *now
    }

    fn get_in_progress_run(
        &self,
        workflow: &str,
        now: &DateTime<Utc>,
    ) -> Result<GithubWorkflowRun> {
        let start_time = Instant::now();
        let interval = Duration::from_millis(500);
        let _span = bar_span!("waiting for workflow to start").entered();
        while start_time.elapsed() < Duration::from_secs(10) {
            let runs = self.list_workflow_runs(&workflow)?.workflow_runs;
            let latest = &runs[0];
            if Self::is_current_run(latest, &now) {
                return Ok(runs[0].clone());
            }
            thread::sleep(interval);
        }

        bail!(
            "timeout while trying to get running action for {:}",
            &workflow
        )
    }
}

#[derive(Debug, Serialize)]
struct GithubWorkflowDispatchBody {
    #[serde(rename = "ref")]
    ref_name: String,

    inputs: HashMap<String, String>,
}

impl GithubWorkflowDispatchBody {
    fn inputs_to_hashmap(inputs: Vec<(String, String)>) -> Result<HashMap<String, String>> {
        let mut map = HashMap::new();
        for (k, v) in &inputs {
            let v = if !v.starts_with("@") {
                v
            } else {
                info!("populating input '{}' with contents from {}", k, &v[1..]);
                &std::fs::read_to_string(&v[1..])
                    .wrap_err(format!("failed to read file {}", &v[1..]))?
            };
            map.insert(k.to_owned(), v.to_owned());
        }
        Ok(map)
    }

    fn new(ref_name: &str, inputs: Vec<(String, String)>) -> Result<Self> {
        let ref_name = ref_name.to_string();
        let inputs =
            Self::inputs_to_hashmap(inputs).wrap_err("failed to generate github workflow body")?;
        Ok(Self { ref_name, inputs })
    }
}

pub fn get_nixpkgs_pr_status(pr_number: i64) -> Result<PrStatus> {
    let repo = GitHubRepo::new("nixos", "nixpkgs")?;
    let pr = repo.get_pr(pr_number)?;
    let status = if !pr.merged {
        PrStatus::NotMerged
    } else {
        PrStatus::Merged {
            nixpkgs_unstable: repo.pr_in_branch("nixos-unstable", &pr)?,
            nixos_unstable_small: repo.pr_in_branch("nixos-unstable-small", &pr)?,
            nixos_unstable: repo.pr_in_branch("nixos-unstable", &pr)?,
        }
    };
    Ok(status)
}

pub fn oizys_gha_run(workflow: &str, ref_name: &str, inputs: Vec<(String, String)>) -> Result<i64> {
    let _span = bar_span!("dispatching github action").entered();
    let repo = GitHubRepo::new("daylinmorgan", "oizys")?;
    let workflow = GitHubRepo::normalize_workflow(&workflow);
    let now: DateTime<Utc> = Utc::now();
    repo.run_action(&workflow, &ref_name, inputs)
        .wrap_err("failed to start action")?;
    let run = repo.get_in_progress_run(&workflow, &now)?;
    Ok(run.id)
}
